### PR TITLE
[SPARK-18209][SQL] More robust view canonicalization without full SQL expansion

### PR DESCRIPTION
**Pull request by jiangxb1987**
## What changes were proposed in this pull request?

Currently we canonicalize the view definition to provide the context for the database as well as star expansion, this is fragile for the following reasons:

1. It is non-trivial to guarantee that the generated SQL is correct without being extremely verbose, given the current set of operators.
2. We need extensive testing for all combination of operators.
3. Whenever we introduce a new logical plan operator, we need to be super careful because it might break SQL generation. 

We use a late binding approach to do view canonicalization, which resolve the dependent views every time the view is used. This PR is expected to achieve the following goals:

1. Views should reflect changes to the underlying tables and intermediate views. For example:
    ```
        CREATE TABLE T1(a int, b string)
        CREATE TABLE T2(a int, b string, c int)
        CREATE VIEW A AS SELECT * FROM T1
        CREATE VIEW B AS SELECT * FROM A
        ALTER VIEW A AS SELECT * FROM T2
        SELECT * FROM B
    ```
will return data from table T2 (Note that we still need to validate the schema).
2. Make views more robust by not using SQL generation to store the view’s logical plan.
3. The views generated by older versions of Spark or HIVE should still work or throw a meaningful error if the view definition is invalid.
4. A view should throw a meaningful error when the view in an invalid state due to an underlying change. Such errors should be thrown as early as possible.

The follow up works will be:

1. Remove the SQL generation (in particular the SQLBuilder class) for Spark SQL operators and the test suites(and also golden files) for them.
2. Disallow cyclic view reference(See [SPARK-18389](https://issues.apache.org/jira/browse/SPARK-18389)).
3. View dependency management similar to Postgres.
4. Underlying table schema change when star is used.
5. Support create permanent views on non-SQL DataFrames.

## How was this patch tested?
Add new test cases in `SQLViewSuite`.

### CODE REVIEW

**Comment by hvanhovell** [comment]
cc @rxin @srinathshankar @nsyca

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveMetastoreCatalog.scala [hunk 1]
```diff
@@ -114,9 +117,26 @@ private[hive] class HiveMetastoreCatalog(sparkSession: SparkSession) extends Log
       alias.map(a => SubqueryAlias(a, qualifiedTable, None)).getOrElse(qualifiedTable)
     } else if (table.tableType == CatalogTableType.VIEW) {
       val viewText = table.viewText.getOrElse(sys.error("Invalid view without text."))
+      val unresolvedPlan = sparkSession.sessionState.sqlParser.parsePlan(viewText).transform {
```
Comment (hvanhovell) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveMetastoreCatalog.scala]: This will probably break with common table expressions. We could just set the current database, have the analyzer solve it, and set the current database back to its old value.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/catalog/interface.scala [hunk 2]
```diff
@@ -151,6 +151,7 @@ case class CatalogTable(
     tableType: CatalogTableType,
     storage: CatalogStorageFormat,
     schema: StructType,
+    originalSchema: Option[StructType] = None,
```
Comment (hvanhovell) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/catalog/interface.scala]: The only difference between originalSchema and schema are the column names, right? If that is the case why not store the `source` column name in the schema field metadata?

**Review by hvanhovell — COMMENTED** [review]
I took a first pass (I haven't looked at the tests yet). This is a good start. I still think that we still need to address a few issues.

It might be nice if you explain in the PR description that we are doing late binding, and how the binding of views work.

**Comment by hvanhovell** [comment]
also cc @gatorsmile

File: sql/core/src/main/scala/org/apache/spark/sql/execution/command/views.scala [hunk 3]
```diff
@@ -207,31 +205,56 @@ case class CreateViewCommand(
...
+        storage = CatalogStorageFormat.empty,
+        schema = aliasedPlan.schema,
+        originalSchema = Some(originalSchema),
+        properties = properties,
+        viewOriginalText = originalText,
+        viewText = Some(viewSQL),
+        currentDatabase = Some(currentDatabase),
+        comment = comment
+      )
+    } else {
```
Comment (nsyca) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/command/views.scala]: I guess the code in the `else` happens when the object is a table, not a view, correct? Please add this to the prologue of the function for clarity.

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/SQLViewSuite.scala [hunk 4]
```diff
@@ -448,19 +476,105 @@ class SQLViewSuite extends QueryTest with SQLTestUtils with TestHiveSingleton {
     }
   }
 
+  test("Using view after change the origin view") {
+    withView("v1", "v2") {
+      sql("CREATE VIEW v1 AS SELECT id FROM jt")
+      sql("CREATE VIEW v2 AS SELECT * FROM v1")
+      withTable("jt2", "jt3") {
+        // Don't change the view schema
+        val df2 = (1 until 10).map(i => i + i).toDF("id")
```
Comment (nsyca) [inline sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/SQLViewSuite.scala]: I suggest to twist this test case a bit.
The first part of the test case has the table jt defined as columns ("j", "i") and then the second part jt2 defined as, what you currently have, columns("i", "j") to test that the code checks the order of the columns correctly.

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/SQLViewSuite.scala [hunk 5]
```diff
@@ -448,19 +476,105 @@ class SQLViewSuite extends QueryTest with SQLTestUtils with TestHiveSingleton {
...
+      sql("CREATE VIEW v2 AS SELECT * FROM v1")
+      // Drop the referenced view
+      sql("DROP VIEW v1")
+      val e = intercept[RuntimeException] {
+        sql("SELECT * FROM v2")
+      }
+      assert(e.getMessage.contains(
+        "Failed to analyze the canonicalized SQL"))
+    }
+  }
```
Comment (nsyca) [inline sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/SQLViewSuite.scala]: I suggest adding a variation of this test case, dropping the underlining table "jt" and v2 should be invalid. This will test the multiple-level expansion of a view. It is also no harm to check that v1 is also invalid in this new test case.

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/SQLViewSuite.scala [hunk 6]
```diff
@@ -448,19 +476,105 @@ class SQLViewSuite extends QueryTest with SQLTestUtils with TestHiveSingleton {
...
+      df.write.format("json").saveAsTable("tab1")
+      withView("v1", "v2") {
+        sql("CREATE VIEW v1 AS SELECT * FROM tab1")
+        sql("CREATE VIEW v2 AS SELECT * FROM v1")
+        // Don't change the table schema
+        val df2 = (1 until 10).map(i => i * i).toDF("i")
+        df2.write.format("json").mode("overwrite").saveAsTable("tab1")
+        // the view v2 should have the same output with the view v1
+        checkAnswer(sql("SELECT * FROM v2"), sql("SELECT * FROM v1"))
+        // the view v2 should have the same output with the table testTable
```
Comment (nsyca) [inline sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/SQLViewSuite.scala]: `testTable` or `tab1`?

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/SQLViewSuite.scala [hunk 7]
```diff
@@ -448,19 +476,105 @@ class SQLViewSuite extends QueryTest with SQLTestUtils with TestHiveSingleton {
...
+        sql("DROP TABLE tab1")
+        val e = intercept[RuntimeException] {
+          sql("SELECT * FROM v1")
+        }
+        assert(e.getMessage.contains(
+          "Failed to analyze the canonicalized SQL"))
+      }
+    }
+  }
+
```
Comment (nsyca) [inline sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/SQLViewSuite.scala]: If you add a new test case to test the multiple-level expansion as I suggested before, this test case is just a duplicate. It may be removed.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/catalog/SessionCatalog.scala [hunk 8]
```diff
@@ -557,9 +557,12 @@ class SessionCatalog(
    * If the relation is a view, the relation will be wrapped in a [[SubqueryAlias]] which will
    * track the name of the view.
    */
-  def lookupRelation(name: TableIdentifier, alias: Option[String] = None): LogicalPlan = {
+  def lookupRelation(
```
Comment (gatorsmile) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/catalog/SessionCatalog.scala]: General suggestion: when you made a change like this, please update the above function descriptions.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala [hunk 9]
```diff
@@ -509,32 +509,42 @@ class Analyzer(
...
     }
 
-    def apply(plan: LogicalPlan): LogicalPlan = plan resolveOperators {
-      case i @ InsertIntoTable(u: UnresolvedRelation, parts, child, _, _) if child.resolved =>
-        i.copy(table = EliminateSubqueryAliases(lookupTableFromCatalog(u)))
-      case u: UnresolvedRelation =>
-        val table = u.tableIdentifier
-        if (table.database.isDefined && conf.runSQLonFile && !catalog.isTemporaryTable(table) &&
+    def apply(plan: LogicalPlan): LogicalPlan = {
+      var currentDatabase = catalog.getCurrentDatabase
```
Comment (gatorsmile) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala]: Why we need to do it here? This should be moved to https://github.com/jiangxb1987/spark/blob/434009ef7fcfc552314dca35dc9395bb98f5e9c6/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala#L539

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala [hunk 10]
```diff
@@ -509,32 +509,42 @@ class Analyzer(
...
-          lookupTableFromCatalog(u)
-        }
+            // If the database part is specified, and we support running SQL directly on files, and
+            // it's not a temporary view, and the table does not exist, then let's just return the
+            // original UnresolvedRelation. It is possible we are matching a query like "select *
+            // from parquet.`/path/to/query`". The plan will get resolved later.
+            // Note that we are testing (!db_exists || !table_exists) because the catalog throws
+            // an exception from tableExists if the database does not exist.
+            u
+          } else {
```
Comment (gatorsmile) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala]: Please roll back all the unnecessary changes between line 522 and line 538

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala [hunk 11]
```diff
@@ -509,32 +509,42 @@ class Analyzer(
...
+            val logicalPlan = lookupTableFromCatalog(u, Some(currentDatabase))
+            currentDatabase = logicalPlan.collectFirst {
```
Comment (gatorsmile) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala]: Please avoid using `var` here.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/catalog/SessionCatalog.scala [hunk 12]
```diff
@@ -557,9 +557,12 @@ class SessionCatalog(
...
+      name: TableIdentifier,
+      alias: Option[String] = None,
+      databaseHint: Option[String] = None): LogicalPlan = {
```
Comment (gatorsmile) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/catalog/SessionCatalog.scala]: Need a document for these parameters.

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveSessionCatalog.scala [hunk 13]
```diff
@@ -55,16 +55,19 @@ private[sql] class HiveSessionCatalog(
...
-    val db = formatDatabaseName(name.database.getOrElse(currentDb))
+    val db = formatDatabaseName(name.database.getOrElse(databaseHint.getOrElse(currentDb)))
     if (db == globalTempViewManager.database) {
       val relationAlias = alias.getOrElse(table)
       globalTempViewManager.get(table).map { viewDef =>
         SubqueryAlias(relationAlias, viewDef, Some(name))
       }.getOrElse(throw new NoSuchTableException(db, table))
     } else if (name.database.isDefined || !tempTables.contains(table)) {
-      val database = name.database.map(formatDatabaseName)
+      val database = Some(db).map(formatDatabaseName)
```
Comment (gatorsmile) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveSessionCatalog.scala]: -> `val database = formatDatabaseName(db)`

Then, you can change the following line to Some(database)

**Comment by gatorsmile** [comment]
@jiangxb1987 We need a low level design about the changes you made here, especially document the issue you hit and the solution you choose.

**Comment by gatorsmile** [comment]
also cc @cloud-fan

**Comment by rxin** [comment]
@jiangxb1987 - please take a look at my comment on the jira ticket: https://issues.apache.org/jira/browse/SPARK-18209

File: sql/core/src/main/scala/org/apache/spark/sql/execution/command/views.scala [hunk 14]
```diff
@@ -207,31 +205,56 @@ case class CreateViewCommand(
...
 
-    // Validate the view SQL - make sure we can parse it and analyze it.
-    // If we cannot analyze the generated query, there is probably a bug in SQL generation.
-    try {
-      sparkSession.sql(viewSQL).queryExecution.assertAnalyzed()
-    } catch {
-      case NonFatal(e) =>
-        throw new RuntimeException(s"Failed to analyze the canonicalized SQL: $viewSQL", e)
-    }
+    if (originalText.isDefined) {
```
Comment (jiangxb1987) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/command/views.scala]: I should add comments over this code. The `originalText` is non-empty only if the command is generated from SQL.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/command/views.scala [hunk 15]
```diff
@@ -207,31 +205,56 @@ case class CreateViewCommand(
...
+      // Validate the view SQL - make sure we can resolve it with currentDatabase.
+      val originalSchema = try {
+        val unresolvedPlan = sparkSession.sessionState.sqlParser.parsePlan(viewSQL)
+        val resolvedPlan = sparkSession.sessionState.analyzer.execute(unresolvedPlan)
+        sparkSession.sessionState.analyzer.checkAnalysis(resolvedPlan)
+
+        resolvedPlan.schema
+      } catch {
+        case NonFatal(e) =>
+          throw new RuntimeException(s"Failed to analyze the canonicalized SQL: $viewSQL", e)
```
Comment (jiangxb1987) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/command/views.scala]: Yeah I agree we should throw a AnalysisException here.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/command/views.scala (hunk 3 above)
Comment (jiangxb1987) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/command/views.scala]: I should add comments over this code. The originalText is non-empty only if the command is generated from SQL.

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveMetastoreCatalog.scala [hunk 16]
```diff
@@ -114,9 +117,26 @@ private[hive] class HiveMetastoreCatalog(sparkSession: SparkSession) extends Log
...
+      val resolvedPlan = try {
+        val resolvedPlan = sparkSession.sessionState.analyzer.execute(unresolvedPlan)
+        sparkSession.sessionState.analyzer.checkAnalysis(resolvedPlan)
+
+        resolvedPlan
+      } catch {
+        case NonFatal(e) =>
+          throw new RuntimeException(s"Failed to analyze the canonicalized SQL: $viewText", e)
+      }
+      val planWithProjection = table.originalSchema.map(withProjection(resolvedPlan, _))
```
Comment (jiangxb1987) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveMetastoreCatalog.scala]: I think so.

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveMetastoreCatalog.scala [hunk 17]
```diff
@@ -114,9 +117,26 @@ private[hive] class HiveMetastoreCatalog(sparkSession: SparkSession) extends Log
...
+        .getOrElse(resolvedPlan)
+
       SubqueryAlias(
         alias.getOrElse(table.identifier.table),
-        sparkSession.sessionState.sqlParser.parsePlan(viewText),
+        aliasColumns(planWithProjection, table.schema.fields),
```
Comment (jiangxb1987) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveMetastoreCatalog.scala]: It will might be a bit more complex, but I think we certainly can do that.

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveMetastoreCatalog.scala [hunk 18]
```diff
@@ -126,6 +146,55 @@ private[hive] class HiveMetastoreCatalog(sparkSession: SparkSession) extends Log
...
+  private def compareStructField(field: StructField, other: StructField): Boolean = {
+    field.name == other.name && field.dataType == other.dataType
+  }
+
+  /**
+   * Aliases the schema of the LogicalPlan to the view attribute names
+   */
+  private def aliasColumns(plan: LogicalPlan, fields: Seq[StructField]): LogicalPlan = {
+    val output = fields.map(field => (field.name, field.getComment))
+    if (plan.output.size != output.size) {
```
Comment (jiangxb1987) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveMetastoreCatalog.scala]: This should not happen. I just want to ensure we are safe here in case the `withProjection` has been modified.

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveSessionCatalog.scala (hunk 13 above)
Comment (jiangxb1987) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveSessionCatalog.scala]: +1

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/SQLViewSuite.scala (hunk 4 above)
Comment (jiangxb1987) [inline sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/SQLViewSuite.scala]: Good point! I'll add the case.

**Comment by jiangxb1987** [comment]
@hvanhovell @nsyca @gatorsmile @rxin Thank you for your suggestions! I will try to make a better approach ASAP! Thank you！

**Comment by jiangxb1987** [comment]
This is closed because we should achieve the final goal by submitting a sequence of smaller PRs.
